Inventory.md

```diff
@@ -0,0 +1,173 @@
+
+### No rollbacks further than "k"
+
+With the Haskell node we will be able to say that old enough state/checkpoints are immutable and can never be rolled back. We might want to change which checkpoints we store in DB. There is no point in keeping multiple checkpoints in the immutable part of the chain.
+
+### Delegation features
+
+We cannot implement delegation features. We could
+1. Keep the API the same, and use dummy implementations internally. E.g return an empty list of stake pools. E.g say that wallets have 0 account balance. I think this should work without any hustle.
+2. Make the wallet modular enough to not have to do the above. I am not sure how difficult this would be.
+
+## Other, fuzzier difficulties of integration
+
+I tried to get a wallet executable working with the Haskell node and discovered new kinds of difficulties.
+
+- Chain parameters are not in the Haskell genesis block. The wallet will need to hard-code or read them from a config file at startup.
+- The genesis block does not have to be treated differently because of above. The jörmungandr wallet currently treats it completely separate.
+- NetworkLayer and `follow` seem like the wrong abstractions to work with both jörmungandre and Haskell.
+- If `follow` is removed from core, `monitorStakePools` breaks (or will have be made jormungandr-specific).
+- Both CLI and api will differ. (E.g. different configuration and cli arguments)
+- Use of BlockHeader as Point could be problematic if inconvenient conversions are needed at inconvenient places.
+
+## Summary Byron
+- Need to deal with absence of `coin` value in transaction inputs (both internally and in the API)
+- Need to deal with of tx fee policy working differently
+- Absence of delegation features requires dummy implementation or re-designed API for Byron
+- There are a few small problems with trivial immediate solutions.
+- There seem to be several fuzzier differences. For instance, the wallet `NetworkLayer` seem to be the wrong abstraction.
+- Some of the fuzzier difficulties might lack cheap temporary solutions. More careful consideration might be useful there, re-evaluating current architecture while keeping future goals in mind.
+
+## Extra: A glimpse into Shelley
+
+- We need to support changing protocol parameters.
+- Delegation features will likely work differently. E.g through imported ledger-rules or by asking for certain ledger state once per epoch. (See Duncan's messages: https://input-output-rnd.slack.com/archives/C819S481Y/p1576776488005100) The very good thing is that we will be able to ask for the ledger state of any point in the unstable chain. In Jörmungandr we could only retrieve the latest stake-distribution and we didn't know what point it belonged to when we got it. This was very tricky to deal with.
```

Comment:
so every time we interact with with node, we will need to ask for protocol parameters. Or haskell node has push mechanism (one can subscribe for change of protocol parameters event) - I am asking as they put huge effort in networking layer so maybe there is a way

Reply:
@paweljakubas the haskell protocols are mostly stateful (i.e., the node remembers where we are at and only changes on explicit requests). 
So, with the haskell nodes, you can easily say:

- Find an intersection between your chain and my local chain
- And now, give me the state of your protocol parameters from this intersection

With jormungandr, finding this intersection is something that WE do within the implementation of the network layer for Jörmungandr. Plus, by the time we've found the intersection, the node has likely moved forward (or worse, switch chain), so by the time you get the actual data in a second HTTP call, you might receive completely outdated data and have to re-sync / find a new intersection and retry. 

To get the sense of it (and how tricky it is in practice), have a look at https://github.com/input-output-hk/cardano-wallet/blob/master/lib/jormungandr/src/Cardano/Wallet/Jormungandr/Network.hs#L383

Fortunately, stateful protocols should make our life MUCH easier because all that logic is now done by the node itself, and our network layer can just blindly follow the node's instructions.

Reply:
> For protocol params, stake pools etc, you have a choice:
> 1. Use the ledger rules (via the lib) to maintain the appropriate subset of the ledger state (everything other than the big things like utxo, reward accounts & stake distribution). With this approach you always have locally the full values and they are automatically in-sync with the block you're up to.
>
> 2. Don't use the ledger rules, and every time you want to find some bit of the ledger state (other than the wallet utxo), you can use the local state query protocol to get reference to a snapshot of the ledger state as of a recent block, and then run queries against it to get what you want. 
>
> [...]
>
> **If it were not for the reward accounts I'd say don't even think about 2 and do everything via 1**.

— [Duncan](https://input-output-rnd.slack.com/archives/C819S481Y/p1576777084005800?thread_ts=1576776488.005100&cid=C819S481Y)

> so every time we interact with with node, we will need to ask for protocol parameters. 

@paweljakubas so no. We will likely know 1) the protocol params at genesis, and 2) when we apply a block we use a function from cardano-ledger to get updated protocol parameters to store in the DB.

From Rodney's [WB-44](https://jira.iohk.io/secure/RapidBoard.jspa?rapidView=46&projectKey=WB&view=planning&selectedIssue=WB-44&issueLimit=100) (Unless I'm missing something subtle this is the same idea):
```Haskell
updateSystemState :: UpdateSystemState -> Block -> UpdateSystemState
```

> And now, give me the state of your protocol parameters from this intersection

@KtorZ I don't believe so. The `LocalStateQuery` protocol will — according to my understanding — allow us to get the full ledger state in the style of `Point -> m FullLedgerState` (but maybe not exactly like that) for points in the **mutable part of the chain**.

https://github.com/input-output-hk/ouroboros-network/blob/9d8493e6d609f879c9ebc684bc5d830f1f146bb6/ouroboros-network/src/Ouroboros/Network/Protocol/LocalStateQuery/Type.hs#L64-L66

Reply:
I will see if I can make this clearer in the document.

Reply:
Hmm. Sweet. So this one would be stateless and allow querying the state from any point in time :thinking: ? Sounds like quite powerful and useful.

Reply:
Yes, it's cool. (Note **mutable part of the chain**, or `within K of the tip` though).

He also said that it is **not complete** and that they want to **work with us to decide what is needed**:

> note that the integration of the local state protocol is not complete. While the pattern of "acquire ref, query, query, release" is set, the exact queries you can do remains to be defined, and we'll want to work with your team to decide what's needed.

Also, I was wondering if we couldn't "jump start" restorations by filtering the full UTxO instead of slow chain sync, but as Duncan pointed out, that doesn't work with sequential wallets, since we don't know which addresses to look for.

Reply:
*cited Duncan's comment directly in the text and added a link to WB-44*